### ISSUE TITLE
Type variable reassignment with particular type constructors causes errors in annotations

### ISSUE THREAD

**Issue by kputnam** [microsoft/pyright#3801]
**Describe the bug**

Pyright reports an incorrect error message under some circumstances.

**To Reproduce**

Write `A = Optional[int]` and enter the same line twice. Any variables annotated with `A` will report the error “Illegal type annotation: variable is not allowed unless it is a type alias”.

This also occurs with `Callable` and `Union`. It doesn't happen with `Set`, `List`, and several other type constructors.

In my IDE, getting type information at the assignment `A = ...` shows the correct type. However variables annotated with type `A` have `Unknown` type, and the type information for the annotation `: A` also shows `A` is `Unknown`.

**Expected behavior**

Except in these few cases (`Optional`, `Union`, `Callable`, perhaps some others), pyright seems to handle type variable reassignments like Python handles variable reassignments. That is, when the type variable is referenced, its most recent value is used up to the point where the reference occurs.

This behavior is what I would expect, but there may be a PEP which specifies different behavior.

**Screenshots or Code**

```python
from typing import Optional

MyUnion = Optional[int]
MyUnion = Optional[int]

ab: MyUnion  = 1
```

This also occurs with `Union`, `Callable`, and perhaps others.

**VS Code extension or command-line**

I’m using pyright 1.1.263 with neovim and nvim-lspconfig. This also happens at the command like when I run `pyright test.py`.

**Comment by erictraut** [comment]
If you want to use `MyUnion` in a type annotation, then it needs to be a legitimate type alias. Prior to [PEP 613](https://peps.python.org/pep-0613/) there was no way to explicitly define a type alias, so type checkers need to use heuristics to distinguish between a type alias and a regular variable. One of the heuristics that pyright (and other type checkers) use is that a type alias can have only one definition. If you define it more than once, it's assumed to be a regular variable, and it cannot be used in a type annotation.
